Expose RoundRobinInetAddressResolver

Spikhalskiy opened the pull request:
Motivation:
Make small refactoring for recently merged PR #5867 to make the code more flexible and expose aggressive round robin as a `NameResolver` too with proper code reuse.

Modifications:
Round robin is a method of hostname resolving - so Round robin related code fully moved to `RoundRobinInetAddressResolver implements NameResolver<InetAddress>`, `RoundRobinInetSocketAddressResolver` is deleted as a separate class, an instance with the same functionality could be created by calling `#asAddressResolver()`.

Result:
New forced Round Robin code exposed not only as an `AddressResolver` but as a `NameResolver` too, more proper code and semantic reusing of `InetNameResolver` and `InetSocketAddressResolver` classes.

normanmaurer commented on resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java:
```diff
@@ -0,0 +1,81 @@
+ */
+@UnstableApi
+public class RoundRobinInetAddressResolver extends InetNameResolver {
+    private NameResolver<InetAddress> nameResolver;
```
final

normanmaurer commented on resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java:
```diff
@@ -0,0 +1,81 @@
+                                // this is to support the round robin distribution
+                                int index =
+                                        (numAddresses == 1)? 0 : ThreadLocalRandom.current().nextInt(numAddresses);
+                                promise.setSuccess(inetAddresses.get(index));
```
just inline the index calculation ?

normanmaurer commented on resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java:
```diff
@@ -0,0 +1,81 @@
+ * A {@link NameResolver} that resolves {@link InetAddress} and force Round Robin by choosing a single address
+ * randomly in {@link #resolve(String)} and {@link #resolve(String, Promise)}
+ * if multiple are returned by the {@link NameResolver}.
+ * Use {@link #asAddressResolver()} to create a {@link java.net.InetSocketAddress} resolver
```
Import `java.net.InetSocketAddress` and so not need to use FQCN here.

normanmaurer commented on resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    protected void doResolve(final String inetHost, final Promise<InetAddress> promise) throws Exception {
+        // hijack the doResolve request, but do a doResolveAll request under the hood
```
missing `.` on EOL.

normanmaurer commented on resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java:
```diff
@@ -0,0 +1,81 @@
+        // Note that InetSocketAddress.getHostName() will never incur a reverse lookup here,
+        // because an unresolved address always has a host name.
+        resolveAll(inetHost)
+                .addListener(new FutureListener<List<InetAddress>>() {
```
merge both above lines

Spikhalskiy commented:
@normanmaurer Addressed your comments. I'm thinking about rework ```ThreadLocalRandom.current().nextInt(numAddresses)``` to just stupid ```private int index = 0; inetAddresses.get(index++/inetAddresses.size())```, even if it's a bit not thread-safe. We don't actually need an expensive random call here. WDYT?

normanmaurer commented:
@Spikhalskiy let us look into this after this is merged..

normanmaurer commented:
Cherry-picked into 4.1 (eb7f8e4dc5802fa7ec1ffbfbc8ac8ef1aacf8918)
